### PR TITLE
Convert to es6 class syntax

### PR DESCRIPTION
**Pull request by s1na**
To ease the transition to typescript, this PR converts `Blockchain` (and `Cache`) to the es6 class syntax and moves `Cache` to a separate file. It adds babel dependencies and configs to build the files to `dist`.

I did a commit to move `index.js` to `src/`, but due to later changes github is not showing it properly... Using the diff of [this commit](https://github.com/ethereumjs/ethereumjs-blockchain/commit/9fd8e7678e1594555f6274ce6a75fb3a4ad169ae) might help with the review process.

I also ran `npm pack` manually and inserted this in `ethereumjs-vm`'s `node_modules` to test, and the tests pass.

Update: Updated the `test` script to build before running tests, and modified the test to import `Blockchain` from `dist/`. I'm not sure what's the best practice here, to test against source itself, or the built version?

### CODE REVIEW

**Comment by holgerd77** [comment]
@s1na Can you rebase this?

On the tests `src` vs `dist` question: on the other libraries we are testing against `src` with some additional integration/consistency tests here and there. I would like to keep it like this for now, otherwise this gets into conflict with linting too much if one wants to add some `console.log` statements e.g. or otherwise do debugging/experiments on the code while testing.

**Comment by s1na** [comment]
Makes sense. Rebased.

File: src/index.js [hunk 1]
```diff
@@ -0,0 +1,1119 @@
...
+      hashToNumber: new Cache({ max: 2048 })
+    }
+    self._initDone = false
+    self._putSemaphore = semaphore(1)
+    self._initLock = new Stoplight()
+    self._init(function (err) {
+      if (err) throw err
+      self._initLock.go()
+    })
+  }
```
Comment (holgerd77) [inline src/index.js]: Ok.

File: src/index.js [hunk 2]
```diff
@@ -0,0 +1,1119 @@
...
+        (cb) => self.db.get(headBlockKey, {
+          keyEncoding: 'binary',
+          valueEncoding: 'binary'
+        }, (err, hash) => {
+          self._headBlock = err ? genesisHash : hash
+          cb()
+        })
+      ], cb)
+    }
+  }
```
Comment (holgerd77) [inline src/index.js]: Ok.

File: src/index.js [hunk 3]
```diff
@@ -0,0 +1,1119 @@
...
+    // ensure init completed
+    self._initLock.await(function runGetHead () {
+      // if the head is not found return the headHeader
+      var hash = self._heads[name] || self._headBlock
+      if (!hash) {
+        return cb(new Error('No head found.'))
+      }
+      self.getBlock(hash, cb)
+    })
+  }
```
Comment (holgerd77) [inline src/index.js]: Ok.

File: src/index.js [hunk 4]
```diff
@@ -0,0 +1,1119 @@
...
+   * @method putBlocks
+   * @param {array} blocks - the blocks to be added to the blockchain
+   * @param {function} cb - a callback function
+   */
+  putBlocks (blocks, cb) {
+    const self = this
+    async.eachSeries(blocks, function (block, done) {
+      self.putBlock(block, done)
+    }, cb)
+  }
```
Comment (holgerd77) [inline src/index.js]: Ok.

File: src/index.js [hunk 5]
```diff
@@ -0,0 +1,1119 @@
...
+        // call fn
+        fn(function () {
+          // leave lock
+          self._putSemaphore.leave()
+          // exit
+          cb.apply(null, arguments)
+        })
+      })
+    }
+  }
```
Comment (holgerd77) [inline src/index.js]: Ok.

File: src/index.js [hunk 6]
```diff
@@ -0,0 +1,1119 @@
...
```
Comment (holgerd77) [inline src/index.js]: Ok.

File: src/index.js [hunk 7]
```diff
@@ -0,0 +1,1119 @@
...
+          key: key,
+          keyEncoding: 'binary',
+          valueEncoding: 'binary',
+          value: value
+        })
+        self._cache.hashToNumber.set(key, value)
+        next()
+      }
+    }
+  }
```
Comment (holgerd77) [inline src/index.js]: Ok.

File: src/index.js [hunk 8]
```diff
@@ -0,0 +1,1119 @@
...
+            if (err) return cb(null, [[], []])
+            cb(null, body)
+          })
+        }
+      }, (err, parts) => {
+        if (err) return cb(err)
+        cb(null, new Block([parts.header].concat(parts.body), {common: self._common}))
+      })
+    }
+  }
```
Comment (holgerd77) [inline src/index.js]: Ok.

File: src/index.js [hunk 9]
```diff
@@ -0,0 +1,1119 @@
...
+
+        mid = Math.floor((min + max) / 2)
+        cb2()
+      })
+    },
+    function onDone (err) {
+      if (err) return cb(err)
+      cb(null, hashes.slice(min))
+    })
+  }
```
Comment (holgerd77) [inline src/index.js]: Ok.

File: src/index.js [hunk 10]
```diff
@@ -0,0 +1,1119 @@
...
+        delete self._staleHeads
+        // set stale headBlock to last previously valid canonical block
+        if (self._staleHeadBlock) {
+          self._headBlock = hash
+          delete self._staleHeadBlock
+        }
+        cb()
+      }
+    })
+  }
```
Comment (holgerd77) [inline src/index.js]: Ok.

File: src/index.js [hunk 11]
```diff
@@ -0,0 +1,1119 @@
...
+
+    // delete all number to hash mappings for deleted block number and above
+    function deleteStaleAssignments (cb2) {
+      if (inCanonical) {
+        self._deleteStaleAssignments(blockNumber, parentHash, dbOps, cb2)
+      } else {
+        cb2()
+      }
+    }
+  }
```
Comment (holgerd77) [inline src/index.js]: Ok.

File: src/index.js [hunk 12]
```diff
@@ -0,0 +1,1119 @@
...
+        })
+      }
+
+      function runFunc (cb3) {
+        var reorg = lastBlock ? lastBlock.hash().equals(block.header.parentHash) : false
+        lastBlock = block
+        func(block, reorg, cb3)
+      }
+    }
+  }
```
Comment (holgerd77) [inline src/index.js]: Ok.

File: src/index.js [hunk 13]
```diff
@@ -0,0 +1,1119 @@
...
+        keyEncoding: 'binary',
+        valueEncoding: 'binary'
+      }, (err, td) => {
+        if (err) return cb(err)
+        self._cache.td.set(key, td)
+        cb(null, new BN(rlp.decode(td)))
+      })
+    })
+  }
+}
```
Comment (holgerd77) [inline src/index.js]: Ok.

**Comment by s1na** [comment]
Yeah, totally agree with both of you! Trying to do it in steps to make reviewing easier.

I was thinking about having pure datastructures for block, blockchain, etc. and adding extra classes which manage state for them, similar to the `ethereumjs-db` idea.

**Pull request merged**